packages/compass-aggregations/src/utils/pipeline-storage.ts

```diff
@@ -57,88 +41,84 @@ function savedPipelineToText(pipeline: StoredPipeline['pipeline']): string {
+  private readonly userData: UserData<typeof PipelineSchema>;
+  constructor(basePath?: string) {
+    this.userData = new UserData(PipelineSchema, {
+      subdir: 'SavedPipelines',
+      basePath,
+    });
+  }
   async loadAll(): Promise<StoredPipeline[]> {
     try {
-      const dir = getDirectory();
-      const files = (await fs.readdir(dir))
-        .filter((file) => file.endsWith('.json'))
-        .map((file) => path.join(dir, file));
-
-      return (
-        await Promise.all(files.map((filePath) => this._loadOne(filePath)))
-      ).filter(Boolean) as StoredPipeline[];
+      const { data } = await this.userData.readAll({
+        ignoreErrors: false,
+        readFileStats: true,
+        mergeStats(input, stats) {
+          return {
+            ...input,
+            lastModified: Number(stats.mtimeMs),
+          };
+        },
+      });
+      return data;
     } catch (e) {
+      debug('Failed to read saved pipelines.', e);
       return [];
     }
   }
 
-  async load(id: string): Promise<StoredPipeline | null> {
-    return this._loadOne(path.join(getDirectory(), `${id}.json`));
-  }
-
-  async _loadOne(filePath: string): Promise<StoredPipeline | null> {
-    try {
-      const [data, stats] = await Promise.all([
-        this._getFileData(filePath),
-        fs.stat(filePath),
-      ]);
-      if (!hasAllRequiredKeys(data)) {
-        return null;
-      }
-      return {
-        ...data,
-        lastModified: stats.mtimeMs,
-        pipelineText: data.pipelineText ?? savedPipelineToText(data.pipeline),
-      };
-    } catch (err) {
-      debug(`Failed to load pipeline ${path.basename(filePath)}`, err);
-      return null;
-    }
-  }
-
-  async _getFileData(filePath: string): Promise<unknown> {
-    const data = await fs.readFile(filePath, ENCODING_UTF8);
-    return JSON.parse(data);
-  }
-
-  /**
-   * Updates attributes of an pipeline.
-   */
-  async updateAttributes(id: string, attributes: Partial<StoredPipeline>) {
-    if (!id) {
-      throw new Error('pipelineId is required');
-    }
-    const dir = getDirectory();
-    // If we are creating a new item and none were created before this directory
-    // might be missing
-    await fs.mkdir(dir, { recursive: true });
-    const filePath = path.join(dir, `${id}.json`);
-    // lastModified is generated on file load, we don't want to store it
-    // eslint-disable-next-line @typescript-eslint/no-unused-vars
-    const { lastModified, ...data } = (await this._loadOne(filePath)) ?? {};
-    const updated = {
-      ...data,
+  async updateAttributes(id: string, attributes: Partial<SaveablePipeline>) {
+    await this.userData.write(id, {
+      ...((await this.userData.readOne(id)) ?? {}),
+      lastModified: Date.now(),
       ...attributes,
-    };
-    await fs.writeFile(
-      filePath,
-      JSON.stringify(updated, null, 2),
-      ENCODING_UTF8
-    );
-    return updated;
+    });
+    return await this.userData.readOne(id);
```

Comment:
We already have this document here, do we need to read it again?

Reply:
The reason we are doing another find here it to get the latest data which is transformed by the schema validator to match the output.

We did talk about it last week and agreed on having this encapsulated within `userdata.write` to return us the latest data. However the userdata does not know the exact structure of the data [can be array (shell-history), object (almost everywhere) or possibly a scaler value] and merging the current value with the update attributes varies and for that reason i'll leave that to the user to handle it.

Reply:
I still think we need to do something for this, we are sacrificing the most common use-case here for one corner case. Feel free to leave it as-is though, I might have time to look into it when tackling switching atlas-service to using user data